OloEngine/src/OloEngine/Renderer/Commands/CommandDispatch.cpp

```
@@ -99,34 +136,261 @@ namespace OloEngine
             api.DisableMultisampling();
     }
 
-    struct CommandDispatchData
+    // Helper: Upload material UBO and bind material textures.
+    // Skips entirely when materialDataIndex matches the last-used index.
+    static void UploadMaterialState(const PODMaterialData& mat, u16 materialDataIndex)
```

Clang (14.0.6)

[warning] function 'UploadMaterialState' has cognitive complexity of 68 (threshold 25)

(readability-function-cognitive-complexity)